### DISCUSSION TITLE
Improve Autocomplete state management

### DISCUSSION THREAD

**Discussion by lucas-santosP** [mantinedev/mantine#4259] [Feature requests]
Is there any plan for improving the Autocomplete state management? The current state that accepts only strings is very limiting.

My main request would be to allow not only a string as the value but also an object. So the autocomplete would need two different states:
- String for the typed value, trigged from `onChange`.
- Object for the selected item from the `data` options, trigged from `onSeleted`.

This is useful in more complex use cases where you have two different valid paths. First where the user can type a free text value, and the second where the user selects from one of the defined options. Something similar to what [MUI Autocomplete](https://mui.com/material-ui/react-autocomplete/) does using the prop `freeSolo`.

**Comment by rtivital** [comment]
Autocomplete supports any kind of data object format in `data` prop – https://mantine.dev/core/autocomplete/#data-prop
You can use `filer` prop to customize filtering logic.

**Reply by lucas-santosP** [reply]
The `data` prop is ok, my request is to change the `value` prop to not only string by also allow object.

**Reply by rtivital** [reply]
You can do that on your side:

```ts
const values = {
  first: { value: 1 },
  second: { value: 2 }
};

const data = ['first', 'second'];
```

It is not planned to introduce this logic into the library since it will make it significantly harder to maintain.

**Reply by lucas-santosP** [reply]
It would really make it more complex to maintain, but would also improve DX and makes the component fits more use cases.

Transforming the data to `<unique key, object>` is a valid idea, but that will not work when the key is not unique. And also when is needed to keep track of both the selected option (object) and the typed value in the input (string)... I tried to do that using `onItemSubmit` prop but without success.

In conclusion, I understand if this is not something you are planning for now because can be complex. But I would like to let my thoughts here for future ideas.
